go-controller/pkg/node/base_node_network_controller_dpu.go

```diff
@@ -14,139 +13,214 @@ import (
+			for nadName := range networkMap {
+				dpuCD := bnnc.podReadyToAddDPU(pod, nadName)
+				if dpuCD != nil {
+					err := bnnc.addDPUPod4Nad(pod, dpuCD, isOvnUpEnabled, netName, nadName, clientSet)
+					if err == nil {
+						servedCache[nadName] = dpuCD
+					}
 				}
-				err = bnnc.addRepPort(pod, vfRepName, podInterfaceInfo, clientSet)
-				if err != nil {
-					klog.Infof("Failed to add rep port, %s. retrying", err)
-					retryPods.Store(pod.UID, true)
-				} else {
-					servedPods.Store(pod.UID, true)
-				}
-			} else {
-				// Handle unscheduled pods later in UpdateFunc
-				retryPods.Store(pod.UID, true)
-				return
 			}
+			bnnc.servedPods.Store(pod.UID, servedCache)
 		},
 		UpdateFunc: func(old, newer interface{}) {
-			pod := newer.(*kapi.Pod)
-			klog.Infof("Update for Pod: %s/%s", pod.ObjectMeta.GetNamespace(), pod.ObjectMeta.GetName())
-			if util.PodWantsHostNetwork(pod) || pod.Status.Phase == kapi.PodRunning {
-				retryPods.Delete(pod.UID)
+			oldPod := old.(*kapi.Pod)
+			newPod := newer.(*kapi.Pod)
+			klog.V(5).Infof("Update for Pod: %s/%s for network %s", newPod.Namespace, newPod.Name, netName)
+			v, ok := bnnc.podNadCache.Load(newPod.UID)
+			if !ok {
+				klog.V(5).Infof("Skipping update for Pod %s/%s as it is not attached to network: %s",
+					newPod.Namespace, newPod.Name, netName)
 				return
 			}
-			_, retry := retryPods.Load(pod.UID)
-			if util.PodScheduled(pod) && retry {
-				if bnnc.name != pod.Spec.NodeName {
-					retryPods.Delete(pod.UID)
-					return
+			networkMap := v.(map[string]*nettypes.NetworkSelectionElement)
+
+			servedCache := map[string]*util.DPUConnectionDetails{}
+			v, ok = bnnc.servedPods.Load(newPod.UID)
+			if ok {
+				servedCache = v.(map[string]*util.DPUConnectionDetails)
+			}
+			for nadName := range networkMap {
+				podDesc := fmt.Sprintf("pod %s/%s for nad %s", newPod.Namespace, newPod.Name, nadName)
+				var oldDPUCD *util.DPUConnectionDetails
+				v, ok := servedCache[nadName]
+				if ok {
+					oldDPUCD = v
 				}
-				vfRepName, err := bnnc.getVfRepName(pod)
-				if err != nil {
-					klog.Infof("Failed to get rep name, %s. retrying", err)
-					return
+				newDPUCD := bnnc.podReadyToAddDPU(newPod, nadName)
+				if !dpuConnectionDetailChanged(oldDPUCD, newDPUCD) {
+					continue
 				}
-				isOvnUpEnabled := atomic.LoadInt32(&bnnc.atomicOvnUpEnabled) > 0
-				podInterfaceInfo, err := cni.PodAnnotation2PodInfo(pod.Annotations, nil, isOvnUpEnabled, string(pod.UID),
-					"", nadName, netName, config.Default.MTU)
-				if err != nil {
-					return
+				if oldDPUCD != nil {
+					// VF already added, but new Pod has changed, we'd need to delete the old VF
+					klog.Infof("Deleting the old VF since either kubelet issued cmdDEL or assigned a new VF or "+
+						"the sandbox id itself changed. Old connection details (%v), New connection details (%v)",
+						oldDPUCD, newDPUCD)
+					err := bnnc.updatePodDPUConnStatusWithRetry(oldPod, nil, nadName)
```

Comment:
Maybe make a function like `addDPUPodForNad()` for this code. E.g. `deleteDPUPod4Nad()`